### PR TITLE
Allow ticks for geoaxes

### PR DESCRIPTION
**Pull request by cvanelteren**
This PR addresses #125. It introduces a test that tests specifically three use cases.
It modifies the code to allow for showing ticks in geoaxis. Currently, the tickers are completely controlled through the backend with small modifications to ensure high quality plots. This PR makes it to add ticks. It does this by manually adding a ticker when new keywords `lontick` or `lattick` is given which can take a bool or numeric as an input. When set to `None` the option is effectively ignored.

The defaults should generate a sensible plot with lat and long lines, but without ticks. Then it specifically tests whether leaving the ticks on on one axis does not hamper the other axes.

image

Edit the parameters added would create a 2^4 test plots in terms of the combinations; I opted for specific tests that are a bit more complex but are quicker to generate while still testing all the properties.

### CODE REVIEW

**Comment by cvanelteren** [comment]
Will fail on added test.

**Comment by cvanelteren** [comment]
Note to self, we may want to add this feature to the docs.

**Comment by cvanelteren** [comment]
Removed some scaling around, but should be good now @beckermr

**Comment by cvanelteren** [comment]
We can do the names as fallback and check with direct computation. I will draft something up.

File: docs/projections.py [hunk 1]
```diff
@@ -344,7 +344,7 @@
 # borders using :ref:`settings <rc_UltraPlot>` like `land`, `landcolor`, `coast`,
 # `coastcolor`, and `coastlinewidth`. Finally, since `ultraplot.axes.GeoAxes.format`
 # calls `ultraplot.axes.Axes.format`, it can be used to add axes titles, a-b-c labels,
-# and figure titles, just like :func:`ultraplot.axes.CartesianAxes.format`.
+# and figure titles, just like :func:`ultraplot.axes.CartesianAxes.format`. UltraPlot also adds the ability to add tick marks for longitude and latitude using the keywords `lontick` and `lattick` for flat projections only. This can enhance contrast and readability under some conditions, e.g. when overlaying contours.
```
Comment (beckermr) [inline docs/projections.py:347]:
```suggestion
# and figure titles, just like :func:`ultraplot.axes.CartesianAxes.format`. UltraPlot also adds the ability to add tick marks for longitude and latitude using the keywords `lontick` and `lattick` for rectilinear projections only. This can enhance contrast and readability under some conditions, e.g. when overlaying contours.
```

File: ultraplot/axes/geo.py [hunk 2]
```diff
@@ -65,6 +71,18 @@
...
+    tick length from rcParams. If `False` or `None`, ticks are not drawn. If a numeric value
+    is provided, ticks are drawn with that length.
+lontick : bool or float, optional
+    Controls tick marks specifically for longitude. Takes precedence over `tick` parameter.
+    If `True`, uses the default tick length. If `False` or `None`, longitude ticks are not drawn.
+    If a numeric value is provided, longitude ticks are drawn with that length in points.
+lattick : bool or float, optional
+    Controls tick marks specifically for latitude. Takes precedence over `tick` parameter.
+    If `True`, uses the default tick length. If `False` or `None`, latitude ticks are not drawn.
+    If a numeric value is provided, latitude ticks are drawn with that length in points.
```
Comment (beckermr) [inline ultraplot/axes/geo.py:85]: The keyword argument names and semantics for these tick marks needs to follow the same API as done in format for `x` and `y` tick marks. See the doc string here: https://github.com/Ultraplot/ultraplot/blob/main/ultraplot/axes/cartesian.py#L38.

File: ultraplot/axes/geo.py [hunk 3]
```diff
@@ -1673,3 +1768,71 @@ def _update_minor_gridlines(self, longrid=None, latgrid=None, nsteps=None):
...
+        # Check if the transformed points are still collinear
+        # Points are collinear if the slopes between consecutive points are equal
+        x0, y0 = transformed_points[0]
+        x1, y1 = transformed_points[1]
+        x2, y2 = transformed_points[2]
+
+        # Calculate slopes
+        if (x1 - x0) == 0 or (x2 - x1) == 0:  # Avoid division by zero
+            # Check if both are vertical lines
+            return (x1 - x0) == 0 and (x2 - x1) == 0
```
Comment (beckermr) [inline ultraplot/axes/geo.py:1807]: These tests for 0 need to use `allclose` as well.

File: ultraplot/axes/geo.py [hunk 4]
```diff
@@ -1673,3 +1768,71 @@ def _update_minor_gridlines(self, longrid=None, latgrid=None, nsteps=None):
 # Apply signature obfuscation after storing previous signature
 GeoAxes._format_signatures[GeoAxes] = inspect.signature(GeoAxes.format)
 GeoAxes.format = docstring._obfuscate_kwargs(GeoAxes.format)
+
+
+def _is_rectilinear_projection(ax):
```
Comment (beckermr) [inline ultraplot/axes/geo.py:1773]: There is some missing coverage in this file. We need to add tests to cover branches.

**Review by beckermr — CHANGES_REQUESTED** [review]
See my comments above.

File: ultraplot/axes/geo.py [hunk 5]
```diff
@@ -65,6 +73,16 @@
 longridminor, latgridminor, gridminor : bool, default: :rc:`gridminor`
     Whether to draw "minor" longitude and latitude lines.
     Use the keyword `gridminor` to toggle both at once.
+lonticks, latticks : locator-spec, optional
+    Used to determine the longitude and latitude axis tick mark positions. Passed
+    to the `~ultraplot.constructor.Locator` constructor. Can be float,
+    list of float, string, or `matplotlib.ticker.Locator` instance.
+    Use `[]`, `'null'`, or `'none'` for no ticks.
```
Comment (beckermr) [inline ultraplot/axes/geo.py:80]: This needs to lonticklen, latticklen, and ticklen. The extra semantics around Boolean values needs to be removed. Ticklen should set the length for both Lon and lat. The principle here is that the api should be identical.

**Review by beckermr — CHANGES_REQUESTED** [review]

File: ultraplot/axes/geo.py [hunk 6]
```diff
@@ -783,13 +783,15 @@ def format(
...
-                f"Projection is not flat ignoring {lonticks=} and {latticks=} settings"
-            )
+        if lonticks or latticks:
+            # Only add warning when ticks are given
+            if _is_rectilinear_projection(self):
+                self._add_geoticks("x", lonticks, tick)
+                self._add_geoticks("y", latticks, tick)
+            else:
+                warnings._warn_ultraplot(
+                    f"Projection is not flat ignoring {lonticks=} and {latticks=} settings"
```
Comment (beckermr) [inline ultraplot/axes/geo.py:793]:
```suggestion
                    f"Projection is not rectilinear. Ignoring {lonticks=} and {latticks=} settings."
```

**Review by beckermr — CHANGES_REQUESTED** [review]

File: ultraplot/axes/geo.py (hunk 4 above)
Comment (cvanelteren) [reply ultraplot/axes/geo.py:1783]: I updated the logic to better reflect the logic within the class. Some of the checks were a bit overkill as they are caught by other functions already. This should satisfy the codecov patches (will see when its updated).  The major change is that some of the checks were a bit overkill since the format function itself sets the projection class and as such we could not check for that with the functions available in pytest, although we could  add tests for private functions (but that would be a bit overkill again).

File: ultraplot/axes/geo.py (hunk 5 above)
Comment (cvanelteren) [reply ultraplot/axes/geo.py:80]: ah I thought you were referring to xtick and ytick

**Comment by cvanelteren** [comment]
todo: rm boolean logic and add unit conversion if string.

**Comment by cvanelteren** [comment]
Looks good now! @beckermr. I will rebase if approved.

**Comment by cvanelteren** [comment]
It is a little tricky to test the untested and partial parts raised by codecov. The main issue is that the code itself does not allow to objects in such a way to these high level function that would create the problems. We could either provide tests for the low level functions, leave them as is, or remove the additional logic that would safeguard against edge cases. I am leaning towards the second option.

**Comment by cvanelteren** [comment]
To add fuel to the fire I also just noticed that there is `_proj_non_rectangular` which also performs a name check. 

I can attempt to modify the projection functions directly using a cartopy class, but the name property is read only -- making testing the latter harder. Basemap has similar saveguards for the names.

File: ultraplot/axes/geo.py [hunk 7]
```diff
@@ -65,6 +73,12 @@
 longridminor, latgridminor, gridminor : bool, default: :rc:`gridminor`
     Whether to draw "minor" longitude and latitude lines.
     Use the keyword `gridminor` to toggle both at once.
+lonticklen, latticklen, ticklen: int, float, bool optional
+
```
Comment (beckermr) [inline ultraplot/axes/geo.py:77]:
```suggestion
```

File: ultraplot/axes/geo.py [hunk 8]
```diff
@@ -761,10 +778,80 @@ def format(
...
             )
+        # Set tick lengths for flat projections
+        if lonticklen or latticklen:
+            # Only add warning when ticks are given
+            if _is_rectilinear_projection(self):
+                self._add_geoticks("x", lonticklen, ticklen)
+                self._add_geoticks("y", latticklen, ticklen)
+            else:
+                warnings._warn_ultraplot(
+                    f"Projection is not flat ignoring {lonticklen=} and {latticklen=} settings"
```
Comment (beckermr) [inline ultraplot/axes/geo.py:789]:
```suggestion
                    f"Projection is not rectilinear. Ignoring {lonticklen=} and {latticklen=} settings."
```

File: ultraplot/axes/geo.py [hunk 9]
```diff
@@ -761,10 +778,80 @@ def format(
...
 
+    def _add_geoticks(self, x_or_y, itick, ticklen):
+        """
+        Add tick marks to the geographic axes.
+
+        Parameters
+        ----------
+        x_or_y : {'x', 'y'}
+            The axis to add ticks to ('x' for longitude, 'y' for latitude).
+        itick, tick : unit-spec, default: :rc:`tick.len`
```
Comment (beckermr) [inline ultraplot/axes/geo.py:803]:
```suggestion
        itick, ticklen : unit-spec, default: :rc:`tick.len`
```

File: ultraplot/axes/geo.py [hunk 10]
```diff
@@ -761,10 +778,80 @@ def format(
...
+            Major tick lengths for the x and y axis.
+            %(units.pt)s
+            Use the keyword `ticklen` to set both at once.
```
Comment (beckermr) [inline ultraplot/axes/geo.py:806]:
```suggestion
            Use the argument `ticklen` to set both at once.
```

File: ultraplot/axes/geo.py [hunk 11]
```diff
@@ -761,10 +778,80 @@ def format(
...
+        Notes
```
Comment (beckermr) [inline ultraplot/axes/geo.py:807]:
```suggestion

        Notes
```

File: ultraplot/axes/geo.py [hunk 12]
```diff
@@ -761,10 +778,80 @@ def format(
...
+        # Skip if no tick size specified
+        if size is None:
+            return
+        size = units(size)
+
+        ax = getattr(self, f"{x_or_y}axis")
+
+        # Get the tick positions based on the locator
+        gl = self.gridlines_major
+        # Note: set_xticks points to a different method than self.[x/y]axis.set_ticks from the mpl backend. For basemap we are adding the ticks to the mpl backend and for cartopy we are simple using their functions by showing the axis.
```
Comment (beckermr) [inline ultraplot/axes/geo.py:823]:
```suggestion
        # Note: set_xticks points to a different method than self.[x/y]axis.set_ticks
        # from the mpl backend. For basemap we are adding the ticks to the mpl backend
        # and for cartopy we are simple using their functions by showing the axis.
```

File: ultraplot/axes/geo.py [hunk 13]
```diff
@@ -761,10 +778,80 @@ def format(
...
+                for object in objects:
+                    # text is wrapped in a list
+                    if isinstance(object, list) and len(object) > 0:
+                        object = object[0]
+                    if isinstance(object, mtext.Text):
+                        object.set_visible(True)
+        else:
+            setattr(gl, f"{x_or_y}padding", 2 * size)
+
+        # Note: set grid_alpha to 0 as it is controlled through the gridlines_major object (which is not the same ticker)
```
Comment (beckermr) [inline ultraplot/axes/geo.py:849]:
```suggestion
        # Note: set grid_alpha to 0 as it is controlled through the gridlines_major
        # object (which is not the same ticker)
```

Comment (beckermr) [inline ultraplot/axes/geo.py:849]: Am I understanding correctly that a user cannot use both ticks and grid lines? We should leave that as configurable by the user and not force them. A nice default is good of course. As the code looks now, I think folks would have to add the gridlines back after adding ticks. Instead, they should only have to make a single call to `.format` that adds ticks and sets the gridline alpha to zero.

Comment (cvanelteren) [reply ultraplot/axes/geo.py]: The ticks are exaggerated in the tests to allow the users to change the size of the ticks

Comment (beckermr) [reply ultraplot/axes/geo.py]: I am confused then. Why is the core setting the grid alpha to zero when adding ticks then?

Comment (cvanelteren) [reply ultraplot/axes/geo.py]: otherwise it doubles the grid lines.

Comment (cvanelteren) [reply ultraplot/axes/geo.py]: Geoaxes are a bit strange. The set tick functions are also no properly overwritten (issue with cartopy) the grid lines in UltraPlot are set manually in a separate function and set tick Params would add it independently

**Review by beckermr — APPROVED** [review]
